tools/x86_64-bootimager/src/lib.rs

```diff
@@ -0,0 +1,297 @@
+            Self::Uefi => "UEFI",
+            Self::Bios => "BIOS",
+        })
+    }
+}
+
+/// Log levels for the `bootloader` crate.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, ValueEnum)]
+#[repr(u8)]
+#[clap(rename_all = "lower")]
+enum BootLogLevel {
+    /// A level lower than all log levels.
+    Off,
+    /// Corresponds to the `Error` log level.
+    Error,
+    /// Corresponds to the `Warn` log level.
+    Warn,
+    /// Corresponds to the `Info` log level.
+    Info,
+    /// Corresponds to the `Debug` log level.
+    Debug,
+    /// Corresponds to the `Trace` log level.
+    Trace,
+}
+
+#[derive(Clone, Debug, Parser)]
+pub struct QemuOptions {
+    /// Path to the QEMU x86_64 executable.
+    ///
+    /// Generally, this does not need to be overridden, unless the QEMU binary
+    /// has a non-standard name or is not on the PATH.
+    #[clap(
+        long,
+        default_value = Self::QEMU_SYSTEM_X86_64,
+        value_hint = ValueHint::FilePath,
+    )]
+    pub qemu_path: Utf8PathBuf,
+
+    /// Extra arguments passed directly to the QEMU command.
+    #[arg(last = true, default_value = "-cpu qemu64 -smp cores=4")]
+    pub qemu_args: Vec<String>,
+}
+
+impl Default for QemuOptions {
+    fn default() -> Self {
+        Self {
+            qemu_path: Utf8PathBuf::from(Self::QEMU_SYSTEM_X86_64),
+            qemu_args: Self::default_args(),
+        }
+    }
+}
+
+impl QemuOptions {
+    const QEMU_SYSTEM_X86_64: &'static str = "qemu-system-x86_64";
+    fn default_args() -> Vec<String> {
+        vec![
+            "-cpu".to_string(),
+            "qemu64".to_string(),
+            "-smp".to_string(),
+            "cores=4".to_string(),
+        ]
+    }
+
+    pub fn run_qemu(
+        &self,
+        bootimage_path: impl AsRef<Utf8Path>,
+        boot_mode: BootMode,
+    ) -> anyhow::Result<()> {
+        let bootimage_path = bootimage_path.as_ref();
+        let QemuOptions {
+            qemu_path,
+            qemu_args,
+        } = self;
+
+        let mut cmd = std::process::Command::new(qemu_path);
+        if !qemu_args.is_empty() {
+            cmd.args(qemu_args.iter());
+        }
+
+        if let BootMode::Uefi = boot_mode {
+            cmd.arg("-bios").arg(ovmf_prebuilt::ovmf_pure_efi());
+        }
+        tracing::info!(
+            ?qemu_args,
+            %qemu_path,
+            %boot_mode,
+            "booting in QEMU: {bootimage_path}"
+        );
+        cmd.arg("-drive")
+            .arg(format!("format=raw,file={bootimage_path}"));
+        let mut qemu = cmd.spawn().context("failed to spawn QEMU child process")?;
+        let status = qemu.wait().context("QEMU child process failed")?;
+
+        if !status.success() {
+            anyhow::bail!("QEMU exited with status: {}", status);
+        }
+
+        Ok(())
+    }
+}
+
+impl fmt::Display for BootLogLevel {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(match self {
+            Self::Off => "off",
```

Comment:
you *could* use `#[derive(derive_more::Display)]`, though that won't give you all-lowercase

Reply:
unfortunately, having it be all lower case is kinda load-bearing because the only reason this *has* a `Display` impl is so that `clap` can show the default value in the CLI help text, and it ought to be formatted the same way that `clap` will parse it there. I had _thought_ that deriving `clap::ValueEnum` would make that work without having to implement `Display` --- which the docs say it does --- but it didn't compile for me, so i just added this. whatever.